File: src/backend/geo_lib/websocket/modules/process_job_module.py
from geo_lib.logging.console import get_websocket_logger
from geo_lib.websocket.base_module import BaseWebSocketModule

logger = get_websocket_logger()

"""
Process job WebSocket module.
Handles process job processing events (status updates, completion, etc.).
"""


class ProcessJobModule(BaseWebSocketModule):
    """WebSocket module for process job processing events."""

    @property
    def module_name(self) -> str:
        return "process_job"

    async def handle_message(self, message_type: str, data: dict) -> None:
        """Handle incoming messages for process job module."""
        if message_type == 'refresh':
            # Process job module doesn't need to send initial state
            # The import queue module handles that
            pass
        else:
            logger.warning(f"Unknown message type for process_job module: {message_type}")

    async def send_initial_state(self) -> None:
        """Process job module doesn't send initial state - import queue module handles that."""
        # Process job module doesn't need to send initial state
        # The import queue module handles the initial state for the UI
        pass

    # WebSocket event handlers for process job events
    async def status_updated(self, event):
        """Handle process job status updates."""
        await self.send_to_client('status_updated', event['data'])

    async def completed(self, event):
        """Handle process job completion."""
        await self.send_to_client('completed', event['data'])

    async def failed(self, event):
        """Handle process job failure."""
        await self.send_to_client('failed', event['data'])

File: src/frontend/src/assets/js/websocket/modules/ProcessJobModule.js
/**
 * Process Job WebSocket module.
 * Handles all process job related realtime events for the import table.
 */

import {BaseModule} from './BaseModule.js';

export class ProcessJobModule extends BaseModule {
    constructor(store) {
        super(store);
        this.moduleName = 'process_job';
    }

    /**
     * Initialize the process job module
     */
    initialize() {
        super.initialize();

        // Handle status updates
        this.subscribe('status_updated', (data) => {
            this.handleStatusUpdate(data);
        });

        // Handle process job completion
        this.subscribe('completed', (data) => {
            // Request refresh of import table to get updated data
            this.socket.requestRefresh('import_queue');
        });

        // Handle process job failure
        this.subscribe('failed', (data) => {
            // Request refresh of import table to get updated data
            this.socket.requestRefresh('import_queue');
        });
    }

    /**
     * Handle status update events
     * @param {Object} data - The status update data
     */
    handleStatusUpdate(data) {
        let updates = {
            processing: data.status === 'processing',
            processing_failed: data.status === 'failed',
            queued: data.status === 'queued'
        };

        // Handle completed status - need to get the actual feature count from the server
        if (data.status === 'completed') {
            updates.processing = false;
            updates.processing_failed = false;
            updates.queued = false;
            // Request a refresh of import table to get the updated item with correct feature count
            this.socket.requestRefresh('import_queue');
            return;
        }

        // For processing status, set feature_count to -1 to indicate processing
        if (data.status === 'processing') {
            updates.feature_count = -1;
            updates.queued = false;
        }

        // For queued status, clear processing flag
        if (data.status === 'queued') {
            updates.processing = false;
        }

        // Update the specific item in the store using import_queue_id
        this.store.dispatch('updateImportTableItem', {
            id: data.import_queue_id,
            updates: updates
        });
    }

    /**
     * Cleanup the process job module
     */
    cleanup() {
        super.cleanup();
    }
}
